### ISSUE TITLE
Revisit opnd_create_base_disp_aarch64

### ISSUE THREAD

**Issue by yury-khrustalev** [DynamoRIO/dynamorio#4530]
This API function for creating base-disp memory reference operands for AArch64

```
opnd_t
opnd_create_base_disp_aarch64(reg_id_t base_reg, reg_id_t index_reg,
                              dr_extend_type_t extend_type, bool scaled, int disp,
                              dr_opnd_flags_t flags, opnd_size_t size);
```

has confusing API especially when comparing it to [Arm A64 specification](https://developer.arm.com/documentation/ddi0487/latest/).

The intent of this ticket is to discuss how this function can be improved. Currently, the problems are:

- This function is essentially a single interface for two different operations (it violates separation of concerns principle).
- It provides too many degrees of freedom and too many possible combinations than it actually should (and most of de jure "supported" pathways are not even checked within the function) (there should one and only one way to do something).
- Parameter `dr_extend_type_t extend_type` implies all values from this enum may be valid, while only two of them actually make sense in this context (well, three if we agree to use `UXTX` for `LSL`).
- Parameter `dr_opnd_flags_t flags` is set to the resulting `opnd`, but is in fact not used (at least not for AArch64), however some of the values of the enum look as if they may affect the operand somehow, so a user would be vainly contemplating which value to choose.
- Parameter `opnd_size_t size` has confusing name and type. It is hard to guess that this is, in fact, a _transfer size_ i.e. the size of the destination operand(s) in the instruction in which this (memref) operand is intended for.

The [code review](https://codereview.appspot.com/308050043) for addition of this function raised some of these concerns, but they had not been addressed.

The decision to store `size` in the operand instead of storing the _shift amount_ was really bad. How can an operand carry a dependency on another operand?

I would like to suggest to replace this function with the following two functions:

```
/**
 * Returns a memory reference operand that refers to a base register X0 - X30 or
 * stack pointer XSP plus an extended and optionally shifted index register.
 *
 * Index register must be either Xm or Wm register. Together with the parameter
 * \p sign_extended it affects the type of extension used for the operand: SXTW
 * and UXTW when using signed and unsigned extension correspondingly with a Wm
 * register and LSL (UXTX) for Xm index register (behaviour for UXTX and SXTX is
 * identical in this context).
 *
 * The \p amount parameter determines optional shift of the extended value of
 * the index register and must be either 0 or log2(sz) where sz is the size of
 * the destination register in bytes (i.e. transfer size: 0 for 8-bit registers,
 * 1 for 16-bit registers, 2 for 32-bit registers, 3 for 64-bit registers, etc).
 *
 * \note AArch64-only.
 */
opnd_t
opnd_create_base_disp_reg(reg_id_t base_reg, reg_id_t index_reg, int amount, bool sign_extended);

/**
 * Returns a memory reference operand that refers to a base register X0 - X30 or
 * stack pointer XSP plus a constant signed displacement.
 *
 * The usable offsets depend on the type of instruction and the size of the
 * destination register (the transfer size).
 *
 * \note AArch64-only.
 */
opnd_t
opnd_create_base_disp_imm(reg_id_t base_reg, reg_id_t index_reg, int offset);
```
Alternatively, a "merged" version of the two functions above may be used (which I don't really like), it would use offset for the displacement when `index_reg = DR_REG_NULL and amount == 0`:

```
opnd_t
opnd_create_base_disp_aarch64(reg_id_t base_reg, reg_id_t index_reg,
                              int amount, bool sign_extended, int offset);
```

But this is only to adhere to the "tradition" introduced by the existing function. I find it really a bad idea to create such facades. Imagine a god-like `opnd_create_opnd` function with a gazillion of parameters.

**Comment by derekbruening** [comment]
+1 to some of the sentiments, but:

> Parameter opnd_size_t size has confusing name and type. It is hard to guess that this is, in fact, a transfer size i.e. the size of the destination operand(s) in the instruction in which this (memref) operand is intended for.

It is not the destination or any other separate operand size: it is the size of this memory operand, i.e., the size of the address range in memory that this operand refers to.

> opnd_create_base_disp_imm(reg_id_t base_reg, reg_id_t index_reg, int offset);

I don't think you meant to have an index register operand?

Note that most existing uses of just base+disp are using `OPND_CREATE_MEM32` and their ilk.  If a C function helper for just base+disp is added I would expect it to be cross-platform as every arch has such operands.

**Comment by yury-khrustalev** [comment]
> the size of the address range in memory that this operand refers to

Well, I described hot it is actually used. I understand that the intended meaning of this parameter was probably different. That's why I said this parameter was confusing.

The same memory reference operand would mean different things when it is used in instructions with different sizes of destination (for load) / source (for stores) operands, so embedding size information in the operand is not strictly correct because this information depends in the outer context (the actual instruction). Technically we should only keep if it's scaled or not, while how much it is scaled by depends on the instruction.

> >  opnd_create_base_disp_imm(reg_id_t base_reg, reg_id_t index_reg, int offset);
> 
> I don't think you meant to have an index register operand?

Yes, there shouldn't be `reg_id_t index_reg` operand here.

> I would expect it to be cross-platform as every arch has such operands.

I would prefer to make the first step with AArch64 function only. Currently API includes several similar functions implemented for different  platforms. Combining them together may lead to another god-like function. Or maybe not. Anyway, it's a much bigger task than what I suggest here. I'm happy to help with AArch64 part.

**Comment by derekbruening** [comment]
> > the size of the address range in memory that this operand refers to
>
> Well, I described hot it is actually used. I understand that the intended meaning of this parameter was probably different. That's why I said this parameter was confusing.

No, it is used in the exact same way as the size of any other operand, such as registers.  I have not heard anyone else think it was confusing.

> The same memory reference operand would mean different things when it is used in instructions with different sizes of destination (for load) / source (for stores) operands

No, it would not.  It means the size of the memory operand.  Other operands have their own sizes and they do not have to match each other.  The memory operand size cannot just be removed: you would break tools and make the IR unusable.

> I would prefer to make the first step with AArch64 function only.

This is not some large or complex or experimental thing where we don't know how it will turn out in the end and we should take incremental steps and only generalize when we know more: this is a very small, well-scoped function.  We know exactly what it does on all platforms, and what it does is pretty much identical everywhere (and in fact the implementation is probably the exact same code for all platforms).  If this function were added but for A64 only, I don't think tool users would use it since it requires an ifdef and duplicated code for general use: they would instead use a cross-platform function like opnd_create_base_disp() with no index register, since an extra operand is simpler than ifdefs and near-duplicate lines.

**Comment by yury-khrustalev** [comment]
> Note that most existing uses of just base+disp are using OPND_CREATE_MEM32 and their ilk.

These are based on `opnd_create_base_disp` and the majority of its usages in DynamoRIO codebase use `DR_REG_NULL` for index register. If a developer wants to create a memory reference operand which would point to a given location, I agree that existing API is enough, but when one needs to create an operand specifically for something like

```
LDR     x1, [sp, w2, uxtw #3]
```
they may benefit from having a specific API function. Currently, one has to write something like

```
opnd_create_base_disp_aarch64(DR_REG_XSP, DR_REG_W2, DR_EXTEND_UXTW, true, 0, 0, OPSZ_8)
```

Note that for the same operand in a different instruction

```
LDR     w1, [sp, w2, uxtw #3]
```

you would use `OPSZ_4`. However no error would be raised if you still used `OPSZ_8` which would essentially create a malformed operand for this instruction.

Use of `DR_EXTEND_UXTW` is redundant here because we use `DR_REG_W2` for index register.

Please also note that from the point of view of writing more cross-platform code, all you need is base register + immediate displacement, however in some cases you may get away with a smaller amount of instructions if you use index registers.

Cross platform API is convenient but it also hinders generating efficient code and makes problematic analysis of platform-specific features. Since we already have functions like `instr_is_sse*` in API which are not marked as platform specific, I think it wouldn't hurt to have and develop API which is specific to other platforms. Users may rely on cross-platform API when they can, and if they need to go deeper, they should always be able to have access to everything supported by their platform of interest. When we have implementation of some function for all supported platforms, someone can oversee all of them and try to encapsulate them under some common and neat interface when possible.
